File: src/FamilyTree.Api/ClientApp/src/types/entity.ts
import type { EntityRefDto } from './entityRef';
import type { LanguageDto } from './language';

export type EntityType = 'Character' | 'Place' | 'Group' | 'Event' | 'Concept';
export type Gender = 'Male' | 'Female' | 'Other' | 'Unknown';

export interface EntityDto {
  id: string;
  name: string;
  type: EntityType;
  worldId: string;
  description?: string;
  firstName?: string;
  lastName?: string;
  middleName?: string;
  maidenName?: string;
  species?: string;
  birthDate?: string;
  birthDateSuffix?: string;
  deathDate?: string;
  deathDateSuffix?: string;
  birthPlaceEntityId?: string;
  birthPlaceEntityName?: string;
  deathPlaceEntityId?: string;
  deathPlaceEntityName?: string;
  residence?: string;
  gender?: Gender;
  biography?: string;
  profilePhotoUrl?: string;
  otherNames?: string;
  position?: string;
  height?: string;
  hairColour?: string;
  parent1Id?: string;
  parent2Id?: string;
  locations: EntityRefDto[];
  affiliations: EntityRefDto[];
  languages: LanguageDto[];
  spouses: EntityRefDto[];
  children: EntityRefDto[];
  createdAt: string;
  updatedAt: string;
}

export interface CreateEntityDto {
  name: string;
  type: EntityType;
  worldId: string;
  description?: string;
  firstName?: string;
  lastName?: string;
  middleName?: string;
  maidenName?: string;
  species?: string;
  birthDate?: string;
  birthDateSuffix?: string;
  deathDate?: string;
  deathDateSuffix?: string;
  residence?: string;
  gender?: Gender;
  biography?: string;
  parent1Id?: string;
  parent2Id?: string;
}

export interface UpdateEntityDto {
  name: string;
  type: EntityType;
  description?: string;
  firstName?: string;
  lastName?: string;
  middleName?: string;
  maidenName?: string;
  species?: string;
  birthDate?: string;
  birthDateSuffix?: string;
  deathDate?: string;
  deathDateSuffix?: string;
  birthPlaceEntityId?: string | null;
  deathPlaceEntityId?: string | null;
  residence?: string;
  gender?: Gender;
  biography?: string;
  profilePhotoUrl?: string;
  otherNames?: string;
  position?: string;
  height?: string;
  hairColour?: string;
  parent1Id?: string | null;
  parent2Id?: string | null;
  locationIds: string[];
  affiliationIds: string[];
  languageIds: string[];
  spouseIds: string[];
  childIds: string[];
}

File: src/FamilyTree.Domain/Entities/Entity.cs
using FamilyTree.Domain.Common;
using FamilyTree.Domain.Enums;

namespace FamilyTree.Domain.Entities;

public class Entity : BaseEntity
{
    public required string Name { get; set; }
    public EntityType Type { get; set; }
    public Guid WorldId { get; set; }
    public string? Description { get; set; }

    // Character-specific (nullable for other types)
    public string? FirstName { get; set; }
    public string? LastName { get; set; }
    public string? MiddleName { get; set; }
    public string? MaidenName { get; set; }
    public string? Species { get; set; }
    public DateOnly? BirthDate { get; set; }
    public string? BirthDateSuffix { get; set; }
    public DateOnly? DeathDate { get; set; }
    public string? DeathDateSuffix { get; set; }
    public Guid? BirthPlaceEntityId { get; set; }
    public Guid? DeathPlaceEntityId { get; set; }
    public string? Residence { get; set; }
    public Gender? Gender { get; set; }
    public string? Biography { get; set; }
    public string? ProfilePhotoUrl { get; set; }
    public string? OtherNames { get; set; }
    public string? Position { get; set; }
    public string? Height { get; set; }
    public string? HairColour { get; set; }

    // Parent references (for family trees)
    public Guid? Parent1Id { get; set; }
    public Guid? Parent2Id { get; set; }

    // Navigation
    public World World { get; set; } = null!;
    public Entity? Parent1 { get; set; }
    public Entity? Parent2 { get; set; }
    public Entity? BirthPlaceEntity { get; set; }
    public Entity? DeathPlaceEntity { get; set; }
    public ICollection<Entity> ChildrenAsParent1 { get; set; } = [];
    public ICollection<Entity> ChildrenAsParent2 { get; set; } = [];
    public ICollection<EntityFamilyTree> EntityFamilyTrees { get; set; } = [];
    public ICollection<Relationship> RelationshipsAsEntity1 { get; set; } = [];
    public ICollection<Relationship> RelationshipsAsEntity2 { get; set; } = [];
    public ICollection<Note> Notes { get; set; } = [];
    public ICollection<EntityLocation> Locations { get; set; } = [];
    public ICollection<EntityLocation> CharactersLocatedHere { get; set; } = [];
    public ICollection<EntityAffiliation> Affiliations { get; set; } = [];
    public ICollection<EntityAffiliation> CharactersAffiliated { get; set; } = [];
    public ICollection<EntityLanguage> Languages { get; set; } = [];
}
